### DISCUSSION TITLE
Is it possible to query qmd render target format at runtime?

### DISCUSSION THREAD

**Discussion by jtlandis** [quarto-dev/quarto-cli#10173] [Q&A]
### Description

I would like to know what the render target of my `.qmd` file is at run time. Specifically to change what is placed in a link for example.

````qmd
---
title: dynamic text
format: 
  html: default
  typst: default
---

```{r}
target <- get_qmd_render_target() # some function like this
words <- sprintf("qmd render target was `%s`", target)
```

Hello! The `r words`.

````

Is this possible?

**Comment by cscheid** [comment]
What do you mean by "render target"?

**Reply by jtlandis** [reply]
I mean `quarto render FILE.qmd --to <format>`, where format is the render target.

**Reply by mcanouil** [reply]
So you want to see in an HTML "html", in a PDF "pdf" and so on?
What exactly are you trying to do?

**Reply by mcanouil** [reply]
I don't really understand the point of this, but you can use Pandoc variable `FORMAT`: https://pandoc.org/lua-filters.html#global-variables

```lua
return {
  ['format'] = function(args, kwargs, meta)
    return tostring(FORMAT)
  end
}
```

````qmd
---
title: "Quarto Playground"
format: pdf
shortcodes:
  - format.lua
---

{{< format >}}
````

**Reply by jtlandis** [reply]
If possible, I'd like to not have to write many inline [content]{.content-visible format-when="FORMAT"} and instead programatically switch the content in some code block.

One example

````qmd

```{r}
format <- get_format()
link <- switch(format, TYPST="https://typst.app/", "https://quarto.org/")
```

[see](`r link`)
````

**Reply by mcanouil** [reply]
Use fenced div, or write a Lua filter that give you an alias.
But you want to have access to this inside R which is not possible. The closest thing you have which is 100% the `knitr` engine and not Quarto is `rmarkdown::metadata` and/or `knitr::is_*_output`

**Reply by cscheid** [reply]
I think `knitr::is_*_output` is the answer here. We don't have an equivalent thing for the Jupyter engine, but we do in `knitr`.

**Reply by mcanouil** [reply]
And if you look at the code you can go to `knitr::opts_knit$get("rmarkdown.pandoc.to")`

**Reply by jtlandis** [reply]
Okay cool! thank you!

**Reply by jtlandis** [reply]
A bit in the same vein, using a lua filter to grab `FORMAT` while within `confluence-html`, I got this result: `"usr/local/quarto-1.4.553/share/filters/customwriter/customwriter.lua`

**Reply by mcanouil** [reply]
Unfortunately, an error line out of context without a small self contained example is not enough to help.

Could you share a **small self-contained "working" (reproducible)** example to work with, _i.e._, a complete Quarto document or a Git repository? The goal is to make it as easy as possible for us to recreate your problem so that we can fix it: please help us help you! Thanks.


---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td></tr>
</table>

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Reply by jtlandis** [reply]
Sure, Please refer to [this repo](https://github.com/jtlandis/quarto-confluence-link/tree/discussion-1073/) which I had made for separate linking issue. The branch `discussion-1073` should have all the reprex needs to show what I mean about  `{{< format >}}` not working for `confluence-html`. 

This example also shows that conditional content is affected by this odd format.

**Reply by mcanouil** [reply]
As stated in https://github.com/quarto-dev/quarto-cli/discussions/10173#discussioncomment-9908445, `FORMAT` is a Pandoc variable.
Confluence is not a Pandoc format. It's a custom format defined by Quarto, thus the weird format name that is passed to Pandoc.

**Reply by jtlandis** [reply]
Also, I am using quarto 1.4.553

**Reply by jtlandis** [reply]
I appreciate the clarity, I don't know too much about Pandoc so thank you.

**Reply by cderv** [reply]
> And if you look at the code you can go to `knitr::opts_knit$get("rmarkdown.pandoc.to")`

The correct way to do this is `knitr::pandoc_to()`. It is easier than remembering the internal option. Usually using the exposed API with R function to get information is recommended. 

`knitr::is_*_output` are wrappers around `pandoc_to()`

> We don't have an equivalent thing for the Jupyter engine, but we do in knitr.

Issue to track for improvement on this is this one

> you can use Pandoc variable FORMAT

It won't completely be equivalent because this will really be the FORMAT as Pandoc can see them. But Quarto has other formats to, including custom format for Pandoc which are Lua filters. 

That is why Quarto needs a way to pass the information of `to:` content to engine if they need it (like confluence, dahsboard, email...). Some context information that only Quarto knows as of today. 

- https://github.com/quarto-dev/quarto-cli/issues/2216
